### PR TITLE
feat: enable caching when using object store

### LINKED ISSUE #918: Enable caching in opendal when using s3/oss storage.
### What type of enhancement is this?

Performance

### What does the enhancement do?

The OpenDAL has a [CacheLayer](https://github.com/datafuselabs/opendal/blob/764e72184e42f5949dd4c8f29770a2e397e89204/src/layers/cache/layer.rs) for caching data content. 

We may enable this layer by default when using s3/OSS storage in the GreptimeDB.

But in future, we can implement more accurate and flexible caching for object storage, especially under distributed and serverless modes.

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by e1ijah1**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Enable cache layer by default when using s3/OSS storage in the GreptimeDB.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
closes #918

### CODE REVIEW

File: src/datanode/src/instance.rs [hunk 1]
```diff
@@ -207,12 +209,17 @@ pub(crate) async fn new_object_store(store_config: &ObjectStoreConfig) -> Result
         ObjectStoreConfig::Oss { .. } => new_oss_object_store(store_config).await,
     };
 
+    let mem_accessor = MemoryBuilder::default().build().unwrap();
```
Comment (killme2008) [inline src/datanode/src/instance.rs:212]: Looks like we enable caching for all storage types, and use memory as cache. I think it's not necessary for local file storage type. Memory cache may consume too much memory, prefer file as caching for s3/OSS. 

May we can add a new configuration `cache_path: Option<String>` to `S3Config` and `OssConfig`, if it is not `None`, we enable the caching for them.

**Comment by killme2008** [comment]
Looks great. It's a pity that opendal doesn't support cache eviction by disk or memory quota, the cached objects may consume too much memory or disk. But I think it's okay right now, we still have to design and impl a more powerful cache layer for object storages.

File: src/datanode/src/instance.rs (hunk 1 above)
Comment (waynexia) [reply src/datanode/src/instance.rs:212]: >prefer file as caching for s3/OSS

+1, caching files on local disk can alleviate lots of remote requests (and maybe latency of reading an object) with lower cost.

**Comment by waynexia** [comment]
I would prefer to leave memory to a more fine-grained cache manager like page cache or query cache. They are much more expensive than disks and may not provide a comparable performance improvement when used to cache the entire file.

File: src/object-store/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+impl ObjectStoreCachePolicy {
+    fn cache_path(&self, path: &str, args: &OpRead) -> String {
+        format!("{}.cache-{}", path, args.range().to_header())
+    }
+}
+
+#[async_trait]
+impl CachePolicy for ObjectStoreCachePolicy {
+    fn on_read(
```
[the remark this answers is not in the document]
Comment (e1ijah1) [reply src/object-store/src/cache.rs:35]: Caching with parameters can generate a lot of fragments, such as file.cache-bytes0-10 and file.cache-bytes1-11. Performance issues may arise if these fragments are generated frequently and cleaned up, and a more granular caching mechanism may be more effective.

[the remark this answers is not in the document]
Comment (killme2008) [reply src/object-store/src/cache.rs:35]: @e1ijah1  Right now, we don't delete objects, but after implementing SST compactions, then there will be some deletions happen. I am concerned that too many cache items may consume too much local disk.  A possible solution is that we use a limited capacity LRU(Least-Recent-Used) cache(for example https://docs.rs/lru/latest/lru/struct.LruCache.html) to track all cached items, and if the LRU cache is full, we can pop the Least-Recent-Used one and delete the object from local disk.

What do you think? @waynexia

[the remark this answers is not in the document]
Comment (waynexia) [reply src/object-store/src/cache.rs:35]: > after implementing SST compactions, then there will be some deletions happen

>A possible solution is that we use a limited capacity LRU(Least-Recent-Used)

Looking good. We can also drop the cache entries when the corresponding resources are removed due to drop table or region migration.

[the remark this answers is not in the document]
Comment (e1ijah1) [reply src/object-store/src/cache.rs]: @killme2008 @waynexia LRU cache has been added. PTAL

File: src/datanode/src/instance.rs [hunk 3]
```diff
@@ -240,7 +241,21 @@ pub(crate) async fn new_oss_object_store(store_config: &ObjectStoreConfig) -> Re
...
+    let mut object_store = ObjectStore::new(accessor);
+    if let Some(path) = oss_config.cache_path.as_ref() {
+        let accessor =
+            FsBuilder::default()
+                .root(path)
+                .build()
+                .with_context(|_| error::InitBackendSnafu {
+                    config: store_config.clone(),
+                })?;
+        let cache_capacity = oss_config.cache_capacity.unwrap();
```
Comment (killme2008) [inline src/datanode/src/instance.rs:253]: We can assume that the user always set `cache_capacity` when setting `cache_path`, I think `unwrap_or(DEFAULT_OBJECT_STORE_CACHE_SIZE)` is better. Define constant `DEFAULT_OBJECT_STORE_CACHE_SIZE`.

File: src/datanode/src/instance.rs [hunk 4]
```diff
@@ -273,7 +288,21 @@ pub(crate) async fn new_s3_object_store(store_config: &ObjectStoreConfig) -> Res
         config: store_config.clone(),
     })?;
 
-    Ok(ObjectStore::new(accessor))
+    let mut object_store = ObjectStore::new(accessor);
+    if let Some(path) = s3_config.cache_path.as_ref() {
+        let accessor =
```
Comment (killme2008) [inline src/datanode/src/instance.rs:293]: Can we extract this code into a function `create_object_store_cache(object_store, path, capacity)`, and reuse it? Looks like it emerges twice.

File: src/object-store/src/cache_policy.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+use opendal::layers::CachePolicy;
+use opendal::raw::output::Reader;
+use opendal::raw::{Accessor, RpDelete, RpRead};
+use opendal::{ErrorKind, OpDelete, OpRead, OpWrite, Result};
+
+struct LruValue;
+
+#[derive(Debug)]
+pub struct LruCachePolicy {
+    lru_cache: Arc<RwLock<LruCache<String, LruValue>>>,
```
Comment (killme2008) [inline src/object-store/src/cache_policy.rs:31]: Looks like it always uses write lock,so we don't need the `RwLock`. I think we can use tokio's Mutex. https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock

File: src/object-store/src/cache_policy.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        inner: Arc<dyn Accessor>,
+        cache: Arc<dyn Accessor>,
+        path: &str,
+        args: OpDelete,
+    ) -> BoxFuture<'static, Result<RpDelete>> {
+        let path = path.to_string();
+        let lru_cache = self.lru_cache.clone();
+        Box::pin(async move {
+            let cache_files: Vec<String> = {
+                let mut guard = lru_cache.write().unwrap();
```
Comment (killme2008) [inline src/object-store/src/cache_policy.rs:106]: Why do we do this iteration? I think just deleting the `cache_path` is enough.

Comment (killme2008) [inline src/object-store/src/cache_policy.rs:106]: Why do we do this iteration? I think just delete the `cache_path` is enough.

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/object-store/src/cache_policy.rs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+use opendal::raw::{Accessor, RpDelete, RpRead};
+use opendal::{ErrorKind, OpDelete, OpRead, OpWrite, Result};
+use tokio::sync::Mutex;
+
+struct LruValue;
+
+#[derive(Debug)]
+pub struct LruCachePolicy {
+    lru_cache: Arc<Mutex<LruCache<String, LruValue>>>,
+}
```
Comment (waynexia) [inline src/object-store/src/cache_policy.rs:33]:
```suggestion
#[derive(Debug)]
pub struct LruCachePolicy {
    lru_cache: Arc<Mutex<LruCache<String, ()>>>,
}
```

You can just put an empty tuple here, like `HashSet<K> = HashMap<K, ()>` in std: https://docs.rs/hashbrown/latest/src/hashbrown/set.rs.html#116-118

**Review by waynexia — APPROVED** [review]
Looking great :+1: Thanks @e1ijah1, I only left a nit-picking style comment.

**Pull request merged**